mephisto/core/task_launcher.py

```diff
@@ -83,13 +96,46 @@ def create_assignments(self) -> None:
                 )
                 self.units.append(Unit(self.db, unit_id))
 
-    def launch_units(self, url: str) -> None:
-        """launch any units registered by this TaskLauncher"""
         for unit in self.units:
+            self.unlaunched_units[unit.db_id] = unit
+
+    def generate_units(self):
+        """ units generator which checks that only 'max_num_concurrent_units' running at the same time,
+        i.e. in the LAUNCHED or ASSIGNED states """
+        while True:
+            for _, launched_unit in self.launched_units.copy().items():
+                if (
+                    launched_unit.get_status() != AssignmentState.LAUNCHED
+                    and launched_unit.get_status() != AssignmentState.ASSIGNED
+                ):
+                    self.num_running_units -= 1
+                    self.launched_units.pop(launched_unit.db_id)
+
+            num_avail_units = self.max_num_concurrent_units - self.num_running_units
+            for i, item in enumerate(self.unlaunched_units.copy().items()):
+                db_id, unit = item
+                if i < num_avail_units:
+                    self.launched_units[unit.db_id] = unit
+                    self.unlaunched_units.pop(unit.db_id)
+                    self.num_running_units += 1
+                    yield unit
+            time.sleep(10)
```

Comment:
Prefer to keep magic numbers in constants near the top of the file.

Reply:
Constant is still here. A definition for `UNIT_GENERATOR_WAIT_SECONDS = 10` would be appropriate near the head of the file.